rpc/RPCTransaction.java

```diff
@@ -135,45 +129,159 @@ public void close() {
         return executeAsync(request, res -> res).get();
     }
 
-    public <T> QueryFuture<T> executeAsync(final TransactionProto.Transaction.Req request, final Function<TransactionProto.Transaction.Res, T> responseReader) {
-        try (ThreadTrace trace = traceOnThread("executeAsync")) {
-            final RPCResponseAccumulator collector = new RPCResponseAccumulator();
-            dispatchRequest(request, collector);
-            return new QueryFuture.Execute<>(collector, responseReader);
+    public <T> QueryFuture<T> executeAsync(TransactionProto.Transaction.Req.Builder request, Function<TransactionProto.Transaction.Res, T> transformResponse) {
+        try (ThreadTrace ignored = traceOnThread("executeAsync")) {
+            final ResponseCollector.Single responseCollector = new ResponseCollector.Single();
+            final UUID requestId = UUID.randomUUID();
+            request.setId(requestId.toString());
+            collectors.put(requestId, responseCollector);
+            return new QueryFuture.Execute<>(request.build(), requestObserver, responseCollector, transformResponse);
         }
     }
 
-    public <T> Stream<T> iterate(final TransactionProto.Transaction.Iter.Req iterRequest, final Function<TransactionProto.Transaction.Iter.Res, T> responseReader) {
-        return iterateAsync(iterRequest, responseReader).get();
+    public <T> Stream<T> iterate(TransactionProto.Transaction.Req.Builder request, Function<TransactionProto.Transaction.Res, T> transformResponse) {
+        return iterateAsync(request, transformResponse).get();
     }
 
-    public <T> QueryFuture<Stream<T>> iterateAsync(final TransactionProto.Transaction.Iter.Req iterRequest, final Function<TransactionProto.Transaction.Iter.Res, T> responseReader) {
-        try (ThreadTrace trace = traceOnThread("iterateAsync")) {
-            final TransactionProto.Transaction.Req request = TransactionProto.Transaction.Req.newBuilder()
-                    .setIterReq(iterRequest).build();
-            final RPCResponseAccumulator collector = new RPCResponseAccumulator();
-            final RPCIterator<T> iterator = new RPCIterator<>(request, responseReader, requestObserver, collector);
-            collectorQueue.add(collector);
-            return new QueryFuture.Stream<>(iterator);
+    public <T> QueryFuture<Stream<T>> iterateAsync(TransactionProto.Transaction.Req.Builder request, Function<TransactionProto.Transaction.Res, T> transformResponse) {
+        try (ThreadTrace ignored = traceOnThread("iterateAsync")) {
+            final ResponseCollector.Multiple responseCollector = new ResponseCollector.Multiple();
+            final UUID requestId = UUID.randomUUID();
+            request.setId(requestId.toString());
+            request.setLatencyMillis(networkLatencyMillis);
+            collectors.put(requestId, responseCollector);
+            return new QueryFuture.Stream<>(request.build(), requestObserver, responseCollector, transformResponse);
```

Comment:
There is an amount of asymmetry between these methods which is far from ideal.

There are a couple of roadblocks here:
- `Future` doesn't contain a `map` method for returning a new `Future` with the result transformed. We could make one, but given that `Future` is a Java-specific thing, I don't think we should overcomplicate our `QueryFuture`. This means we cannot simply take `transformResponse` out of `executeAsync`.
- Adding `transformResponse` to `execute` would make calls to `execute` sharply less readable. Take one example:
```
return rpcTransaction.execute(request).getThingRes();
```
becomes
```
return rpcTransaction.execute(request, TransactionProto.Transaction.Res::getThingRes);
```
Adding `transformResponse` to `execute` is surely less hideous than taking it out of `executeAsync` and creating `QueryFuture.map`, but given that there are many calls to `execute` and only one definition of `execute`, I would rather that the calls looked nice.